File: app/src/main/java/com/wzl/loudnessplayer/audio/R128Meter.kt
package com.wzl.loudnessplayer.audio

import kotlin.math.PI
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToInt
import kotlin.math.tan

/**
 * Streaming approximation of ITU-R BS.1770 / EBU R128 integrated loudness.
 *
 * It applies the standard K-weighting filters, measures overlapping 400 ms
 * blocks every 100 ms, then applies the -70 LUFS absolute gate and the
 * relative -10 LU gate. MP3 is normally mono or stereo; additional channels
 * use equal weighting as a safe fallback.
 */
class R128Meter(
    sampleRate: Int,
    private val channelCount: Int,
) {
    init {
        require(sampleRate > 0)
        require(channelCount > 0)
    }

    private val filters = Array(channelCount) { KWeightingFilter(sampleRate.toDouble()) }
    private val windowFrames = max(1, (sampleRate * BLOCK_SECONDS).roundToInt())
    private val hopFrames = max(1, (sampleRate * HOP_SECONDS).roundToInt())
    private val energyWindow = DoubleArray(windowFrames)
    private val blockEnergies = mutableListOf<Double>()

    private var windowPosition = 0
    private var framesInWindow = 0
    private var framesSinceBlock = 0
    private var energySum = 0.0
    private var peak = 0.0

    fun addFrame(samples: DoubleArray) {
        require(samples.size >= channelCount)

        var frameEnergy = 0.0
        for (channel in 0 until channelCount) {
            val sample = samples[channel].coerceIn(-1.0, 1.0)
            peak = max(peak, kotlin.math.abs(sample))
            val weighted = filters[channel].process(sample)
            frameEnergy += weighted * weighted
        }

        if (framesInWindow == windowFrames) {
            energySum -= energyWindow[windowPosition]
        } else {
            framesInWindow += 1
        }

        energyWindow[windowPosition] = frameEnergy
        energySum += frameEnergy
        windowPosition = (windowPosition + 1) % windowFrames

        if (framesInWindow == windowFrames) {
            if (blockEnergies.isEmpty()) {
                emitBlock()
            } else {
                framesSinceBlock += 1
                if (framesSinceBlock >= hopFrames) {
                    emitBlock()
                }
            }
        }
    }

    fun result(): LoudnessResult? {
        val absoluteGated = blockEnergies.filter { energy ->
            energy > 0.0 && loudnessForEnergy(energy) >= ABSOLUTE_GATE_LUFS
        }
        if (absoluteGated.isEmpty()) return null

        val relativeGate =
            loudnessForEnergy(absoluteGated.average()) + RELATIVE_GATE_LU
        val finalGate = max(ABSOLUTE_GATE_LUFS, relativeGate)
        val relativeGated = absoluteGated.filter { loudnessForEnergy(it) >= finalGate }
        if (relativeGated.isEmpty()) return null

        val loudness = loudnessForEnergy(relativeGated.average())
        val peakDbfs = if (peak > 0.0) 20.0 * log10(peak) else Double.NEGATIVE_INFINITY
        return LoudnessResult(
            integratedLufs = loudness,
            samplePeakDbfs = peakDbfs,
        )
    }

    private fun emitBlock() {
        blockEnergies += energySum / windowFrames
        framesSinceBlock = 0
    }

    private fun loudnessForEnergy(energy: Double): Double =
        LUFS_OFFSET + 10.0 * log10(energy)

    data class LoudnessResult(
        val integratedLufs: Double,
        val samplePeakDbfs: Double,
    )

    private class KWeightingFilter(sampleRate: Double) {
        private val shelf = highShelf(sampleRate)
        private val highPass = highPass(sampleRate)

        fun process(sample: Double): Double = highPass.process(shelf.process(sample))

        private companion object {
            fun highShelf(sampleRate: Double): Biquad {
                val frequency = 1681.974450955533
                val gainDb = 3.999843853973347
                val quality = 0.7071752369554196
                val k = tan(PI * frequency / sampleRate)
                val vh = 10.0.pow(gainDb / 20.0)
                val vb = vh.pow(0.4996667741545416)
                val a0 = 1.0 + k / quality + k * k
                return Biquad(
                    b0 = (vh + vb * k / quality + k * k) / a0,
                    b1 = 2.0 * (k * k - vh) / a0,
                    b2 = (vh - vb * k / quality + k * k) / a0,
                    a1 = 2.0 * (k * k - 1.0) / a0,
                    a2 = (1.0 - k / quality + k * k) / a0,
                )
            }

            fun highPass(sampleRate: Double): Biquad {
                val frequency = 38.13547087602444
                val quality = 0.5003270373238773
                val k = tan(PI * frequency / sampleRate)
                val a0 = 1.0 + k / quality + k * k
                return Biquad(
                    b0 = 1.0 / a0,
                    b1 = -2.0 / a0,
                    b2 = 1.0 / a0,
                    a1 = 2.0 * (k * k - 1.0) / a0,
                    a2 = (1.0 - k / quality + k * k) / a0,
                )
            }
        }
    }

    private class Biquad(
        private val b0: Double,
        private val b1: Double,
        private val b2: Double,
        private val a1: Double,
        private val a2: Double,
    ) {
        private var x1 = 0.0
        private var x2 = 0.0
        private var y1 = 0.0
        private var y2 = 0.0

        fun process(input: Double): Double {
            val output = b0 * input + b1 * x1 + b2 * x2 - a1 * y1 - a2 * y2
            x2 = x1
            x1 = input
            y2 = y1
            y1 = output
            return output
        }
    }

    private companion object {
        const val BLOCK_SECONDS = 0.4
        const val HOP_SECONDS = 0.1
        const val ABSOLUTE_GATE_LUFS = -70.0
        const val RELATIVE_GATE_LU = -10.0
        const val LUFS_OFFSET = -0.691
    }
}

File: ios/LoudnessPlayerIOS/Sources/Audio/R128Meter.swift
import Foundation

struct LoudnessResult: Equatable, Sendable {
    let integratedLUFS: Double
    let samplePeak: Double
}

enum R128MeterError: Error, Equatable {
    case invalidFormat
    case insufficientAudio
    case belowAbsoluteGate
}

struct R128Meter: Sendable {
    private let sampleRate: Double
    private let channels: Int
    private let windowFrames: Int
    private let stepFrames: Int
    private var filters: [KWeightingFilter]
    private var energyWindow: [Double]
    private var windowIndex = 0
    private var populatedFrames = 0
    private var framesSinceBlock = 0
    private var rollingEnergy = 0.0
    private var blockEnergies: [Double] = []
    private var peak = 0.0
    private var hasInvalidFormat = false

    init(sampleRate: Double, channels: Int) {
        self.sampleRate = sampleRate
        self.channels = channels
        windowFrames = max(1, Int((sampleRate * 0.4).rounded()))
        stepFrames = max(1, Int((sampleRate * 0.1).rounded()))
        filters = (0..<max(0, channels)).map { _ in KWeightingFilter(sampleRate: sampleRate) }
        energyWindow = [Double](repeating: 0, count: max(1, windowFrames))
        hasInvalidFormat = sampleRate <= 0 || channels <= 0
    }

    mutating func consume(_ chunk: PCMChunk) {
        guard !hasInvalidFormat,
              chunk.channels == channels,
              abs(chunk.sampleRate - sampleRate) < 0.5,
              chunk.samples.count >= Int(chunk.frameCount) * channels else {
            hasInvalidFormat = true
            return
        }

        for frame in 0..<Int(chunk.frameCount) {
            var frameEnergy = 0.0
            for channel in 0..<channels {
                let raw = Double(chunk.samples[frame * channels + channel])
                peak = max(peak, abs(raw))
                let weighted = filters[channel].process(raw)
                frameEnergy += weighted * weighted
            }

            if populatedFrames == windowFrames {
                rollingEnergy -= energyWindow[windowIndex]
            } else {
                populatedFrames += 1
            }
            energyWindow[windowIndex] = frameEnergy
            rollingEnergy += frameEnergy
            windowIndex = (windowIndex + 1) % windowFrames
            framesSinceBlock += 1

            if populatedFrames == windowFrames,
               (blockEnergies.isEmpty || framesSinceBlock >= stepFrames) {
                blockEnergies.append(max(0, rollingEnergy / Double(windowFrames)))
                framesSinceBlock = 0
            }
        }
    }

    func result() throws -> LoudnessResult {
        if hasInvalidFormat { throw R128MeterError.invalidFormat }
        if blockEnergies.isEmpty { throw R128MeterError.insufficientAudio }
        let absolutelyGated = blockEnergies.filter { loudness(of: $0) >= -70 }
        guard !absolutelyGated.isEmpty else { throw R128MeterError.belowAbsoluteGate }
        let ungatedEnergy = absolutelyGated.reduce(0, +) / Double(absolutelyGated.count)
        let relativeGate = loudness(of: ungatedEnergy) - 10
        let relativelyGated = absolutelyGated.filter { loudness(of: $0) >= relativeGate }
        guard !relativelyGated.isEmpty else { throw R128MeterError.belowAbsoluteGate }
        let integratedEnergy = relativelyGated.reduce(0, +) / Double(relativelyGated.count)
        return LoudnessResult(integratedLUFS: loudness(of: integratedEnergy), samplePeak: peak)
    }

    private func loudness(of energy: Double) -> Double {
        guard energy > 0 else { return -.infinity }
        return -0.691 + 10 * log10(energy)
    }
}

private struct KWeightingFilter: Sendable {
    private var shelf: Biquad
    private var highPass: Biquad

    init(sampleRate: Double) {
        shelf = .highShelf(
            sampleRate: sampleRate, frequency: 1_681.974_450_955_533,
            gainDB: 3.999_843_853_973_347, q: 0.707_175_236_955_419_6
        )
        highPass = .highPass(
            sampleRate: sampleRate, frequency: 38.135_470_876_024_44,
            q: 0.500_327_037_323_877_3
        )
    }

    mutating func process(_ sample: Double) -> Double {
        highPass.process(shelf.process(sample))
    }
}

private struct Biquad: Sendable {
    let b0: Double
    let b1: Double
    let b2: Double
    let a1: Double
    let a2: Double
    private var x1 = 0.0
    private var x2 = 0.0
    private var y1 = 0.0
    private var y2 = 0.0

    static func highShelf(
        sampleRate: Double, frequency: Double, gainDB: Double, q: Double
    ) -> Biquad {
        let k = tan(Double.pi * frequency / sampleRate)
        let vh = pow(10, gainDB / 20)
        let vb = pow(vh, 0.499_666_774_154_541_6)
        let denominator = 1 + k / q + k * k
        return Biquad(
            b0: (vh + vb * k / q + k * k) / denominator,
            b1: 2 * (k * k - vh) / denominator,
            b2: (vh - vb * k / q + k * k) / denominator,
            a1: 2 * (k * k - 1) / denominator,
            a2: (1 - k / q + k * k) / denominator
        )
    }

    static func highPass(sampleRate: Double, frequency: Double, q: Double) -> Biquad {
        let k = tan(Double.pi * frequency / sampleRate)
        let denominator = 1 + k / q + k * k
        return Biquad(
            b0: 1 / denominator, b1: -2 / denominator, b2: 1 / denominator,
            a1: 2 * (k * k - 1) / denominator,
            a2: (1 - k / q + k * k) / denominator
        )
    }

    mutating func process(_ input: Double) -> Double {
        let output = b0 * input + b1 * x1 + b2 * x2 - a1 * y1 - a2 * y2
        x2 = x1; x1 = input; y2 = y1; y1 = output
        return output
    }
}
